ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -0,0 +1,176 @@
+data Transition = Event ProgressEvent | StopProgress
+
+updateState :: IO () -> Transition -> State -> IO State
+updateState _      _                    Stopped     = pure Stopped
+updateState start (Event KickStarted)   NotStarted  = Running <$> async start
+updateState start (Event KickStarted)   Completed   = Running <$> async start
+updateState start (Event KickStarted)   (Running a) = cancel a >> Running <$> async start
+updateState _     (Event KickCompleted) (Running a) = cancel a $> Completed
+updateState _     (Event KickCompleted) st          = pure st
+updateState _     StopProgress          (Running a) = cancel a $> Stopped
+updateState _     StopProgress          st          = pure st
+
+-- | Data structure to track progress across the project
+data InProgress = InProgress
+    { todo    :: !Int  -- ^ Number of files to do
+    , done    :: !Int  -- ^ Number of files done
+    , current :: !(HMap.HashMap NormalizedFilePath Int)
+    }
+
+recordProgress :: NormalizedFilePath -> (Int -> Int) -> InProgress -> InProgress
+recordProgress file shift InProgress{..} = case HMap.alterF alter file current of
+    ((prev, new), m') ->
+        let todo' = if isJust prev then todo else todo + 1
+            done' = if new == 0 then done+1 else done
+        in InProgress todo' done' m'
+  where
+    alter x = let x' = maybe (shift 0) shift x in ((x,x'), Just x')
+
+-- | A 'ProgressReporting' that enqueues Begin and End notifications in a new
+--   thread, with a grace period (nothing will be sent if 'KickCompleted' arrives
+--   before the end of the grace period).
+delayedProgressReporting
+  :: Seconds  -- ^ Grace period before starting
+  -> Seconds  -- ^ sampling delay
+  -> Maybe (LSP.LanguageContextEnv c)
+  -> ProgressReportingStyle
+  -> IO ProgressReporting
+delayedProgressReporting before after lspEnv optProgressStyle = do
+    inProgressVar <- newVar $ InProgress 0 0 mempty
+    progressState <- newVar NotStarted
+    let progressUpdate event = modifyVar_ progressState $ updateState (mRunLspT lspEnv $ lspShakeProgress inProgressVar) (Event event)
+        progressStop   = modifyVar_ progressState $ updateState (mRunLspT lspEnv $ lspShakeProgress inProgressVar) StopProgress
+        inProgress :: NormalizedFilePath -> Action a -> Action a
+        inProgress = withProgressVar inProgressVar
+    return ProgressReporting{..}
+    where
+        lspShakeProgress inProgress = do
+            -- first sleep a bit, so we only show progress messages if it's going to take
+            -- a "noticable amount of time" (we often expect a thread kill to arrive before the sleep finishes)
+            liftIO $ sleep before
+            u <- ProgressTextToken . T.pack . show . hashUnique <$> liftIO newUnique
+
+            b <- liftIO newBarrier
+            void $ LSP.sendRequest LSP.SWindowWorkDoneProgressCreate
+                LSP.WorkDoneProgressCreateParams { _token = u } $ liftIO . signalBarrier b
+            ready <- liftIO $ waitBarrier b
```

Comment:
If we send a WorkDoneProgressCreate, do we need to clean it up ever, or does it just expire on its own?

Reply:
As far as  I know it doesn't require any clean up